### PR TITLE
[Whisper] Fix timestamp processor

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Mostly adds condition when looking in the past and in the futur based on timing information. 
Fixes the tests

### CODE REVIEW

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -101,75 +100,81 @@ def _find_timestamp_sequence(sequences, tokenizer, feature_extractor, max_source
...
         if seq_idx != 0:
             time -= stride_left + stride_right
             offset = int((time / feature_extractor.sampling_rate) / time_precision)
-            timestamp_tokens = np.where(sequence >= timestamp_begin)[0][1::2]
-            if len(timestamp_tokens) >= 1:
-                # if a big chunk lenght is used, we need to check all of the previous items
+            overlap_time = int((stride_left / feature_extractor.sampling_rate) / time_precision)
+            # relevant timestamps are the ones that are in the overlapping part
+            relevant_timestamp = np.where(sequence[consecutive] >= timestamp_begin + overlap_time)[0]
+            if relevant_timestamp.shape[0] > 0:
```
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:113]: We only take the next sequences that are timely relevant

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -101,75 +100,81 @@ def _find_timestamp_sequence(sequences, tokenizer, feature_extractor, max_source
...
+                relevant_timestamp = (
+                    consecutive[relevant_timestamp[0] - 1] if relevant_timestamp[0] > 0 else consecutive[0]
+                )
+                # if a big stride is used, we need to check some of the previous items for the best overlap
                 best_match = 0
                 sliced_sequence = []
                 for idx, previous_sequence in enumerate(reversed(items)):
                     previous_tokens = previous_sequence[1:-1]
+                    if previous_sequence[0] < (timestamp_begin + offset - overlap_time) and idx != 0:
+                        break  # the previous sequence is too far in the past
```
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:123]: If we are looking too far in the past, we will go through the whole sequence. This cuts based on the possible overlapping timing data

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -101,75 +100,81 @@ def _find_timestamp_sequence(sequences, tokenizer, feature_extractor, max_source
...
-                # single timestamp at the end means no speech after the last timestamp.
-                last_idx = np.argwhere(sequence == timestamps[-1])[0][0]
-                sliced_sequence = sequence[: last_idx + 1]
-                duration = sliced_sequence[-1] - sliced_sequence[0]
-                # We need to discard the previous timing information
-                sliced_sequence[0] = items[-1][-1]
-                sliced_sequence[-1] = items[-1][-1] + duration
-                items.append(sliced_sequence)
-        # The beginning time of the next chunk
+                last_slice = current_slice
```
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:176]: This was redondant and repetitive, fixed now

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 4]
```diff
@@ -344,7 +344,7 @@ def test_find_longest_common_subsequence(self):
             },
         )
         merge = _find_timestamp_sequence(
-            [[previous_sequence, (3000, 0, 0)], [next_sequences_1, (3000, 750, 0)]],
+            [[previous_sequence, (480_000, 0, 0)], [next_sequences_1, (480_000, 120_000, 0)]],
```
Comment (ArthurZucker) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:347]: The actual timing information were not really used at this point so the tests were passing but they should not have been 😉

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 5]
```diff
@@ -101,75 +100,81 @@ def _find_timestamp_sequence(sequences, tokenizer, feature_extractor, max_source
...
+        last_timestamp = np.where(timestamp_tokens)[0][-1]
+        consecutive = np.append(consecutive, last_timestamp) if last_timestamp not in consecutive else consecutive
         if seq_idx != 0:
             time -= stride_left + stride_right
             offset = int((time / feature_extractor.sampling_rate) / time_precision)
-            timestamp_tokens = np.where(sequence >= timestamp_begin)[0][1::2]
-            if len(timestamp_tokens) >= 1:
-                # if a big chunk lenght is used, we need to check all of the previous items
+            overlap_time = int((stride_left / feature_extractor.sampling_rate) / time_precision)
+            # relevant timestamps are the ones that are in the overlapping part
```
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:111]:
```suggestion
            # relevant timestamps are in the overlapping part
```

**Comment by ArthurZucker** [comment]
Tested with a concatenated librispeech, took 393.1348168849945 seconds, and with a WER 0.030776774096215136

**Review by sgugger — APPROVED** [review]
LGTM but let's make sure @Narsil agrees as well before merging.

**Review by Narsil — APPROVED** [review]
LGTM.

Does this fix: https://github.com/huggingface/transformers/issues/21179 ?

It does not seem to, right ?

**Comment by ArthurZucker** [comment]
0hhh right let me add this in! Since it is related

**Pull request merged**